### PR TITLE
Add unit tests for `fisher` function in `utils`

### PR DESCRIPTION
**Pull request by Sjoerd1993**
Fixes issue #53 

In this current PR the following unit tests are performed for the `fisher` function:

 - `test_fisher_information_values` tests for a known case whether the calculated values are still correct
 - `test_fisher_shape` Tests whether the shape of the fisher matrix remains correct when using different model parameters
 - `test_fisher_diagonal_positive` tests for varying sets of q-values whether the diagonal elements are greater than zero. As far as I understood, these diagonal elements represent the FI for the individual parameter on this row, and should never be zero.
 
Also, the triple assignments in `fisher` function have been removed, and the associated docstring now indicates the list type for `qs` and `counts` (a list of list),

 I refactored the general structure of the unit test quite a bit since this morning, moving the `get_fisher_information` outside of the testing Class, as to allow to test for different cases. Given it's a cheap calculation, it is probably still worth it to just calculate g in each test.

There's probably a few more tests that could be added as well, such as the importance scaling and the gradient calculations. But I figured those may be added in a separate PR:

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,8 +1,7 @@
 **__pycache__
+.*
```
Comment (jfkcooper) [inline .gitignore:2]: I would have done this in a separate PR, but it is fine here now

File: hogben/tests/test_fisher.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    substrate = SLD(2.047, name='Substrate')(thick=0, rough=2)
+    layer1.thick.bounds = (40, 70)
+    layer2.thick.bounds = (100, 180)
+    layer1.rough.bounds = (6, 10)
+    layer2.rough.bounds = (1, 3)
+    sample = air | layer1 | layer2 | substrate
+
+    # Define model
+    angle_times = [(0.7, 100, 5), (2.0, 100, 20)]  # (Angle, Points, Time)
+    model, _ = simulate(sample, angle_times)
```
Comment (jfkcooper) [inline hogben/tests/test_fisher.py:26]: this would probably be better just calling refnx directly:
` model = refnx.reflect.ReflectModel(sample, scale=scale, bkg=bkg, dq=dq)`

File: hogben/tests/test_fisher.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+import pytest
+
+from hogben.simulate import simulate
+from hogben.utils import fisher
+from refnx.reflect import SLD
+from unittest.mock import patch
+
+
+@pytest.fixture
+def model_fixture():
```
Comment (jfkcooper) [inline hogben/tests/test_fisher.py:11]: call this `refnx_model` or similar, you have already made it clear that it is a fixture

File: hogben/tests/test_fisher.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
+import numpy as np
+import pytest
+
+from hogben.simulate import simulate
+from hogben.utils import fisher
+from refnx.reflect import SLD
+from unittest.mock import patch
+
+
+@pytest.fixture
```
Comment (jfkcooper) [inline hogben/tests/test_fisher.py:10]: I would also add a refl1d model fixture as we should make sure we are testing for both types of input

File: hogben/tests/test_fisher.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    Mocks the reflectivity values in the calculation, values are changed between each call
+    to mimic a changing parameter in the model. Value should always be between 0 and 1.
+    """
+    r = [i / data_points for i in range(data_points)]  # Create reflectivity values from 0 to approx. 1
+    while True:
+        r = [abs(value - 0.43) for value in r]  # Change reflectivity after each call
+        yield r
+
+
+class Test_Fisher():
```
Comment (jfkcooper) [inline hogben/tests/test_fisher.py:59]: check out https://github.com/ferdinandvanwyk/python_project_example for how the tests might want to be laid out, as it stands, you don't need these in a class

File: hogben/tests/test_fisher.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        Tests that all the calculated Fisher information matrix returns the expected values
+        for a given set of parameters
+        """
+        g = get_fisher_information(model_fixture)
+        expected_fisher = [[2.59014803e+04, 2.07211842e+05, 4.60470761e+02, 6.90706141e+01],
+                           [2.07211842e+05, 1.65769474e+06, 3.68376609e+03, 5.52564913e+02],
+                           [4.60470761e+02, 3.68376609e+03, 8.18614686e+00, 1.22792203e+00],
+                           [6.90706141e+01, 5.52564913e+02, 1.22792203e+00, 1.84188304e-01]]
+        np.testing.assert_allclose(g, expected_fisher, rtol=1e-08, atol=0)
+
```
Comment (jfkcooper) [inline hogben/tests/test_fisher.py:71]: can you add a test to show that changing the step in the fisher calculation does not change the information (by much at least)

File: hogben/tests/test_fisher.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    mock_reflectivity.side_effect = get_mock_reflectivity(len(qs[0]))
+    return fisher(qs, xi, counts, [model_fixture])
+
+
+def get_mock_reflectivity(data_points):
+    """
+    Mocks the reflectivity values in the calculation, values are changed between each call
+    to mimic a changing parameter in the model. Value should always be between 0 and 1.
+    """
+    r = [i / data_points for i in range(data_points)]  # Create reflectivity values from 0 to approx. 1
```
Comment (jfkcooper) [inline hogben/tests/test_fisher.py:53]: this could be np.arange I think?

File: hogben/tests/test_fisher.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        yield r
+
+
+class Test_Fisher():
+    def test_fisher_information_values(self, model_fixture):
+        """
+        Tests that all the calculated Fisher information matrix returns the expected values
+        for a given set of parameters
+        """
+        g = get_fisher_information(model_fixture)
```
Comment (jfkcooper) [inline hogben/tests/test_fisher.py:65]: One of the reasons we would want to mock things in here would be to get an answer which we can make sense of, e.g. analytically calculate without too much trouble - e.g. remove the reflectivity and model from the calculation. I would modify this test so we can do this, and include another test where we are not mocking, but have the whole workflow tested, with the numbers which it spits out.

**Review by jfkcooper — CHANGES_REQUESTED** [review]
Quite a few comments here, please message if any are not clear, or if you want further info on how to implement them.

File: hogben/tests/test_fisher.py (hunk 4 above)
Comment (Sjoerd1993) [reply hogben/tests/test_fisher.py:10]: I was actually working on this, just before I read this review. I was testing the coverage using `pytest --cov-report html hogben/tests/ --cov=hogben`, which gives very useful overviews when entire lines are missed. From which I noticed the ref1d case was missed, as well as the no data case.

File: hogben/tests/test_fisher.py (hunk 8 above)
Comment (Sjoerd1993) [reply hogben/tests/test_fisher.py:65]: I've got it to work with a separate test that runs through the whole workflow (for both models), as well as with a test that uses mocked data for both the model and reflectivity (no physical calculations/simulations involved).
It does require me however to refactor `fisher` quite a bit, so that I can mock the gradient matrix J, and the bounds. This way, the need to call an actual model is completely removed. See latest commit. 

I'll trigger a request for review when this is ready. Current code runs, but won't win any beauty awards. Any comments on splitting up `fisher` function in particular are welcome either way. 

Tomorrow I will clean up the testing code, address the remaining comments and create some proper test cases.

File: hogben/tests/test_fisher.py (hunk 2 above)
Comment (Sjoerd1993) [reply hogben/tests/test_fisher.py]: I tried to do this for the refl1d fixture as well, but refl1d is a bit less convenient to work with and requires some processing which is all handled in `simulate`. 

I could move that logic to the refl1d fixture as well, so that a call to simulate with a blank `_` parameter is not needed, but I still used the `simulate` method for the refl1d fixture to keep it cleaner. I'm happy to just repeat that logic in the fixture though if that's preferred.

**Comment by Sjoerd1993** [comment]
With regard to the latest PR, there's some bigger changes:

- Two seperate methods can be called to either generate a proper calculation of the Fisher matrix, or two create one with mocked parameters and reflectivity (bypassing any models, simulations and reflectivity calculations). 
- The fisher matrix is tested against known values using the refl1d and refnx model 
- A test is added to check that an empty array of qs gives the proper empty matrix
- A test is added to check that the Fisher matrix values remain (mostly) the same when changing step size
- A mock of Xi (the list of parameters) has been created, as opposed to mocking the gradient calculation. The thinking being that the gradient calculation itself should be tested either way, and also that it's nice to be able to give actual dummy values for each parameter in the tests.

- Some refactoring has been done in fisher function. This was mostly needed to facilitate the patch in `get_fisher_mock`

File: hogben/tests/test_fisher.py [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+
+@pytest.mark.parametrize('qs',
+                         (np.arange(0.001, 1.0, 0.25),
+                          np.arange(0.001, 1.0, 0.10),
+                          np.arange(0.001, 1.0, 0.05),
+                          np.arange(0.001, 1.0, 0.01)))
+def test_fisher_diagonal_positive(qs):
+    """Tests whether the diagonal values in the Fisher information matrix
+     are positively valued"""
+    qs = [0.1, 0.2, 0.4, 0.6, 0.8]
```
Comment (jfkcooper) [inline hogben/tests/test_fisher.py:208]: this overwrites the `qs` which you paramterize every time

File: hogben/tests/test_fisher.py [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+    g = get_fisher_mock(qs=[qs])
+    np.testing.assert_array_less(np.zeros(len(g)), np.diag(g))
+
+
+@pytest.mark.parametrize('xi',
+                         (MockXi([]),
+                          MockXi([8]),
+                          MockXi([8, 2]),
+                          MockXi([8, 2, 60]),
+                          MockXi([8, 2, 60, 150])))
```
Comment (jfkcooper) [inline hogben/tests/test_fisher.py:218]: this can definitely be called in a nicer way, but see my comments about making it a fixture to be called

File: hogben/tests/test_fisher.py [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+def test_fisher_no_data(xi):
+    """Tests whether a model with zero data points properly returns a
+    zero array"""
+    qs = []
+    g = get_fisher_mock(qs=[qs], xi=xi)
```
Comment (jfkcooper) [inline hogben/tests/test_fisher.py:223]: just call `g = get_fisher_mock(qs=[], xi=xi)` here, but also it should be testing `fisher` not the mocked version I think

File: hogben/utils.py [hunk 12]
```diff
@@ -138,49 +138,29 @@ def __corner(self, results):
...
         return fig
 
-def fisher(qs, xi, counts, models, step=0.005):
+def fisher(qs: list, xi: list, counts: list, models: list, step: float=0.005):
     """Calculates the Fisher information matrix for multiple `models`
        containing parameters `xi`.
 
     Args:
-        qs (list): Q points for each model.
+        qs (list[list]): Q points for each model.
```
Comment (jfkcooper) [inline hogben/utils.py:146]: I have recently gone from typing in the docstring to typing in the arguments list. I think we should only do it in the args, and not have reprtition (also as it stands these two disagree with eachother!)

**Review by jfkcooper — CHANGES_REQUESTED** [review]
a couple more to come

**Comment by Sjoerd1993** [comment]
The refl1d model fixture would need to be refactored a bit once the `simulate` class is refactored (as far as I can tell, that's the only part in `test_fisher.py` that actually depends on `simulate`), but I thought it would make most sense to do that once that refactor is actually in.  

I could move the logic that it borrows from `simulate` to a helper function in `test_fisher.py` otherwise, but that would basically mean copying that current function to here, which would not be pretty and add several extra dependencies from the `simulate` class.

File: hogben/tests/test_fisher.py [hunk 13]
```diff
@@ -0,0 +1,337 @@
...
+
+@pytest.mark.parametrize('model_class', ("mock_refl1d_model",
+                                         "mock_refnx_model"))
+@pytest.mark.parametrize('model_params', (1, 2, 3, 4))
+def test_fisher_no_data(model_params, model_class, request):
+    """Tests whether a model with zero data points properly returns an empty
+    matrix of the correct shape"""
+    model = request.getfixturevalue(model_class)
+    xi = model.xi[:model_params]
+    qs = []
```
Comment (jfkcooper) [inline hogben/tests/test_fisher.py:294]: just write these directly in to `fisher` e.g. `qs=[]`

File: hogben/tests/test_fisher.py [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+    zero array"""
+    model = request.getfixturevalue(model_class)
+    mock_reflectivity.side_effect = get_mock_reflectivity()
+    g = get_fisher_information([model], xi=[])
+    np.testing.assert_equal(g.shape, (0, 0))
+
+
+@pytest.mark.parametrize('model_class', ("refnx_model",
+                                         "refl1d_model"))
+def test_multiple_identical_models(model_class, request):
```
Comment (jfkcooper) [inline hogben/tests/test_fisher.py:313]: I think this could have a slightly more descriptive name, e.g. what we are testing is that the FIM scales when we repeat a "measurement"

**Review by jfkcooper — CHANGES_REQUESTED** [review]
Some more comments, getting quite close now.

File: hogben/tests/test_fisher.py [hunk 15]
```diff
@@ -0,0 +1,337 @@
...
+        count_list = np.ones(len(qs[0])) * 100
+        counts = [count_list for _ in models]  # One list of counts per model
+    if xi is None:
+        xi = []  # Concatenate all xi's to a single list of parameters.
+        for model in models:
+            xi += model.xi
+    return fisher(qs, xi, counts, models, step)
+
+
+def get_reflectivity_given_datapoints(data_points):
```
[the remark this answers is not in the document]
Comment (Sjoerd1993) [reply hogben/tests/test_fisher.py:136]: Unfortunately this does not work directly. The `side_effect` attribute needs to be called on a generator function, which `np.random.rand(len(qs))` is not. I can set the `return_value` using `mock_reflectivity.return_value = np.random.rand(len(qs))`, but that executes the function only once and then returns the same value on subsequent calls.

A thing I can do, is to simplify `get_reflectivity_given_datapoints` significantly using the suggested comment here:
```
def get_random_reflectivity(data_points):
    while True:
        yield np.random.rand(data_points)
```

Which yields a different array of reflectivity points each time, and indeed passes testing because it doesn't actually use the exact values of the returned numbers.

[the remark this answers is not in the document]
Comment (jfkcooper) [reply hogben/tests/test_fisher.py:136]: I'm pretty sure the generator doesn't need to be its own function

[the remark this answers is not in the document]
Comment (Sjoerd1993) [reply hogben/tests/test_fisher.py:136]: You can indeed define a generator function in a list-comprehension style using `()` brackets, I was not aware of that.
Do mind that a generator function slowly gets emptied, and you cannot create an infinite loop. 

The suggested function just yields the random values from a single array (generated by `np.random.rand(len(qs))`), instead of yielding the list itself. But what does work is `mock_reflectivity.side_effect = (np.random.rand(len(qs)) for _ in range(9))`. I'm not sure if I'm the biggest fan of the unused index to let it loop a predefined amount of 9 times, but I couldn't find a clean solution here.

The 9 in this range is because reflectivity is called eight times. Twice per time the unit test is performed. I guess it should work with 8 as well, but it doesn't, perhaps it yields the first value in the generator function on assignment.

File: hogben/tests/test_fisher.py [hunk 16]
```diff
@@ -0,0 +1,337 @@
...
+@patch('hogben.utils.reflectivity')
+@pytest.mark.parametrize('model_class', ("mock_refl1d_model",
+                                         "mock_refnx_model"))
+@pytest.mark.parametrize('qs',
+                         (np.arange(0.001, 1.0, 0.25),
+                          np.arange(0.001, 1.0, 0.10),
+                          np.arange(0.001, 1.0, 0.05),
+                          np.arange(0.001, 1.0, 0.01)))
+def test_fisher_diagonal_positive(mock_reflectivity, qs, model_class, request):
+    """Tests whether the diagonal values in the Fisher information matrix
```
[the remark this answers is not in the document]
Comment (Sjoerd1993) [reply hogben/tests/test_fisher.py:278]: My thinking was that the limit goes to zero, but never approaches it. A zero value of the FIM would basically resemble infinite certainty about a parameter, right? Which would be nonphysical, it would require an infinite shift in reflectivity for the step in the gradient matrix.  Which is why I thought I'd test for positives, instead of non-negatives.

File: hogben/tests/test_fisher.py [hunk 17]
```diff
@@ -0,0 +1,337 @@
...
+        qs: The Q points for each model.
+        step: step size to take when calculating gradient.
+    Returns:
+        numpy.ndarray: Fisher information matrix for given models and data.
+    """
+    # Provide default values for qs, counts and xi
+    if qs is None:
+        # Add one list of qs per model
+        qs = [np.array([0.1, 0.2, 0.4, 0.6, 0.8]) for _ in models]
+    if counts is None:
```
[the remark this answers is not in the document]
Comment (Sjoerd1993) [reply hogben/tests/test_fisher.py:125]: `Counts` does change for different user tests where the total amount of q points change. But I refactored this anyway, and deleted this function.

File: hogben/tests/test_fisher.py [hunk 18]
```diff
@@ -0,0 +1,337 @@
...
+        Mock(spec=bumps.parameter.Parameter, value=value,
+             bounds=Mock(limits=[lb, ub]))
+        for value, lb, ub in parameter_values
+    ]
+    model = Mock(spec=refl1d.experiment.Experiment, xi=parameters)
+    model.xi = parameters
+    return model
+
+
+def get_fisher_information(models: list[Union['refnx.reflect.ReflectModel',
```
[the remark this answers is not in the document]
Comment (Sjoerd1993) [reply hogben/tests/test_fisher.py:101]: The logic for this function was to create some sort of constructor function, where only the arguments need to be supplied that are not equal to the defaults. Which can be convenient, as some cases require a bit of logic that was handled centrally (e.g. the dynamic scaling of counts with the amount of qs).

Either way, I've refactored this a bit, and instead it just calls `fisher()` directly on each unit test, like suggested here. If an input argument requires any sort of different logic, it's handled at the corresponding unit test instead.

File: hogben/tests/test_fisher.py [hunk 19]
```diff
@@ -0,0 +1,337 @@
...
+        [-4.23068571e-07, -1.84017555e-07, 4.51142562e-08, 8.21397190e-08],
+        [-6.80596824e-07, -2.83513452e-07, 8.21397190e-08, 1.62625881e-07]
+    ]
+    np.testing.assert_allclose(g, expected_fisher, rtol=1e-08)
+
+
+@patch('hogben.utils.reflectivity')
+@pytest.mark.parametrize('model_class', ("mock_refl1d_model",
+                                         "mock_refnx_model"))
+def test_fisher_no_importance_correct_values(mock_reflectivity, model_class,
```
[the remark this answers is not in the document]
Comment (Sjoerd1993) [reply hogben/tests/test_fisher.py:195]: Perhaps gotten a bit too verbose in the docstring, as I included the values for the matrices as well...

File: hogben/tests/test_fisher.py (hunk 14 above)
Comment (Sjoerd1993) [reply hogben/tests/test_fisher.py:313]: Renamed it to `test_fisher_doubling_with_two_identical_models`, as it tests that the FIM has double values when using two identical models.

File: hogben/tests/test_fisher.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (jfkcooper) [reply hogben/tests/test_fisher.py:289]: If your parameterisation of the problem has been done in such a way that none of your measured points depend on the parameters you have chosen then it is completely fine to have zero information about those parameters. think about points taken below the critical edge: they have no information about the SiOx layer thickness...

File: hogben/tests/test_fisher.py [hunk 20]
```diff
@@ -0,0 +1,354 @@
...
+        Mock(spec=bumps.parameter.Parameter, value=value,
+             bounds=Mock(limits=[lb, ub]))
+        for value, lb, ub in parameter_values
+    ]
+    model = Mock(spec=refl1d.experiment.Experiment, xi=parameters)
+    model.xi = parameters
+    return model
+
+
+def get_random_reflectivity(data_points):
```
Comment (jfkcooper) [inline hogben/tests/test_fisher.py:104]: This now doesn't get used, so delete

File: hogben/tests/test_fisher.py [hunk 21]
```diff
@@ -0,0 +1,354 @@
...
+        [-0.25, -0.1 , -0.5 ],
+        [-0.25, -0.1 , -0.5 ],
+        [-0.25, -0.1 , -0.5 ]
+
+    M is given by by:
+    M =  [ 100.,    0.,    0.,    0.,    0.],
+         [   0.,  200.,    0.,    0.,    0.],
+         [   0.,    0.,  250.,    0.,    0.],
+         [   0.,    0.,    0.,  500.,    0.],
+         [   0.,    0.,    0.,    0., 1000.]
```
Comment (jfkcooper) [inline hogben/tests/test_fisher.py:179]: I actually like this docstring, verbose though it may be

File: hogben/tests/test_fisher.py [hunk 22]
```diff
@@ -0,0 +1,224 @@
...
+
+@pytest.mark.parametrize('step', (0.01, 0.0075, 0.0025, 0.001, 0.0001))
+def test_fisher_consistent_steps_refl1d(step, refl1d_model):
+    """
+    Tests whether the Fisher information remains mostly consistent when
+    changing step size using the refl1d model
+    """
+    g_reference = get_fisher_information(refl1d_model, step=0.005)
+    g_compare = get_fisher_information(refl1d_model, step=step)
+    np.testing.assert_allclose(g_reference, g_compare, rtol=1e-02, atol=0)
```
[the remark this answers is not in the document]
Comment (jfkcooper) [reply hogben/tests/test_fisher.py:258]: one more `atol` to remove

File: hogben/tests/test_fisher.py [hunk 23]
```diff
@@ -0,0 +1,354 @@
...
+                          np.arange(0.001, 1.0, 0.01)))
+def test_fisher_diagonal_non_negative(mock_reflectivity, qs, model_class,
+                                   request):
+    """Tests whether the diagonal values in the Fisher information matrix
+     are all zero or greater"""
+    model = request.getfixturevalue(model_class)
+    mock_reflectivity.side_effect = (np.random.rand(len(qs)) for _ in range(9))
+    counts = [np.ones(len(qs)) * 100]
+    g = fisher([qs], model.xi, counts, [model])
+    assert min(np.diag(g)) >= 0 # Check if smallest value is at least zero
```
Comment (jfkcooper) [inline hogben/tests/test_fisher.py:296]: I see what you are doing, but probably it is better to do:
`assert np.all(np.diag(g)) >= 0`

**Review by jfkcooper — CHANGES_REQUESTED** [review]
Very very minor! sorry

**Review by jfkcooper — APPROVED** [review]
Good job!
